rfcs/remote-schema-permissions.md

```diff
@@ -0,0 +1,244 @@
+
+**Interfaces**
+
+```
+allowed_interfaces:
+- name: Shape
+  fields: [height, width]
+```
+
+Interface type can also specify a subset of fields (similar to object types)
+
+**Unions**
+
+```
+allowed_unions:
+- name: SearchResult
+  objects: [Person]
+```
+
+Union types can specify the subset of list of object types.
+
+
+**Enums**
+
+```
+allowed_enums:
+- name: Direction
+  values: [North, South]
+```
+
+Enum types can specify a subset of values for the enum.
+
+**Input Objects**
+
+```
+allowed_input_objects:
+- name: Point2D
+  input_fields: [x]
+```
+
+Input objects can specify a subset of input fields.
+
+#### Wildcard syntax
+
+Since specifying each and every type might be tedious (especially for types that may not really require masking like scalars), we can give provide a wildcard syntax:
+
+```
+allowed_objects:
+- name: Query
+  fields: [hello, user]
+- name: User
+  fields: "*"
+allowed_custom_scalars: "*"
+```
+
+In short, any list reference can be replaced by `*` to mean "all valid values" for that list. For example,
+
+1. For an object type, like "User", `fields` can be `*`
+2. Even the root level definition can take `*` like `allowed_custom_scalars: *`
+
+
+## Role-based argument presets
+
+The above API only illustrates schema masking. We need to incorporate argument presets as well to provide "filtered" results. The API described in this section is the final proposed API.
+
+Each field in GraphQL can take arguments as well. Arguments must belong to input types i.e Scalar, Enum, or Input Object (or list of these).
+
+Role-based argument presets can basically inject values from session variables or static values during execution. The presetted argument should also get removed from the exposed role-based schema.
+
+The proposed API:
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields: 
+      - name: hello
+      - name: user
+        argument_presets:
+        - name: "id"
+          value:
+            from_session_variable: "x-hasura-user-id"
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+Argument presets can also go arbitrarily deep if the input type is Input Object (this is very similar to how we define the mapping in remote relationships):
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields:
+        - name: hello
+        - name: user
+          argument_presets:
+          - name: "where"
+            input_field:
+            - name: "id"
+              input_field:
+              - name: "_eq"
+                value:
+                  from_session_variable: "x-hasura-user-id"
+          - name: "limit"
+            value:
+              static: 1
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
```

Comment:
This (just tackling Objects + fields, and argument presets) seems enough to me; is there a reason we need to deal with interfaces, etc?

Reply:
Not really, it is just for completeness. We can scope it to just object types.